### PR TITLE
[webview_flutter] get/set a custom User Agent

### PR DESCRIPTION
**Pull request by LuisThein**
## Description

Based on this pull request https://github.com/flutter/plugins/pull/968 that was closed, I added the functionality to set a custom User Agent and get the current User Agent in the current WebView version.

## Related Issues

https://github.com/flutter/flutter/issues/28256

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by LuisThein** [comment]
Updated emails CLA.

**Comment by tp** [comment]
Thank you @LuisThein, we're already making use of this as we had the same requirement to change the user agent :)

File: packages/webview_flutter/lib/webview_flutter.dart [hunk 1]
```diff
@@ -255,6 +256,18 @@ class WebView extends StatefulWidget {
   /// By default `debuggingEnabled` is false.
   final bool debuggingEnabled;
 
+  /// The User_Agent used for WebView requests.
+  ///
+  /// By default `userAgent` is null.
+  /// If the `userAgent` is null, the default User Agent from the Android / iOS WebView is used.
```
Comment (amirh) [inline packages/webview_flutter/lib/webview_flutter.dart]: nit: When null the platform's webview default is used for the `User-Agent` header.

File: packages/webview_flutter/lib/webview_flutter.dart [hunk 2]
```diff
@@ -255,6 +256,18 @@ class WebView extends StatefulWidget {
   /// By default `debuggingEnabled` is false.
   final bool debuggingEnabled;
 
+  /// The User_Agent used for WebView requests.
+  ///
+  /// By default `userAgent` is null.
```
Comment (amirh) [inline packages/webview_flutter/lib/webview_flutter.dart]: nit: move this to the bottom on its own paragraph.

File: packages/webview_flutter/lib/webview_flutter.dart [hunk 3]
```diff
@@ -255,6 +256,18 @@ class WebView extends StatefulWidget {
...
+  /// If the `userAgent` is null, the default User Agent from the Android / iOS WebView is used.
+  /// WebView User Agent for Android (https://developer.chrome.com/multidevice/user-agent)
+  ///
+  /// When the WebView is rebuild with a new User Agent, the page reloads and the request uses the new User Agent.
```
Comment (amirh) [inline packages/webview_flutter/lib/webview_flutter.dart]: s/User Agent/user agent/

File: packages/webview_flutter/lib/webview_flutter.dart [hunk 4]
```diff
@@ -255,6 +256,18 @@ class WebView extends StatefulWidget {
   /// By default `debuggingEnabled` is false.
   final bool debuggingEnabled;
 
+  /// The User_Agent used for WebView requests.
+  ///
+  /// By default `userAgent` is null.
+  /// If the `userAgent` is null, the default User Agent from the Android / iOS WebView is used.
+  /// WebView User Agent for Android (https://developer.chrome.com/multidevice/user-agent)
```
Comment (amirh) [inline packages/webview_flutter/lib/webview_flutter.dart]: s/WebView/[WebView]/

File: packages/webview_flutter/lib/webview_flutter.dart (hunk 3 above)
Comment (amirh) [inline packages/webview_flutter/lib/webview_flutter.dart]: s/rebuild/rebuilt/

File: packages/webview_flutter/lib/webview_flutter.dart [hunk 5]
```diff
@@ -255,6 +256,18 @@ class WebView extends StatefulWidget {
...
+  ///
+  /// When the WebView is rebuild with a new User Agent, the page reloads and the request uses the new User Agent.
+  ///
+  /// When [WebViewController.goBack] is called after changing the User Agent, the old User Agent is used.
```
Comment (amirh) [inline packages/webview_flutter/lib/webview_flutter.dart]: Suggestion:

When [WebViewController.goBack] is called after changing `userAgent` the previous `userAgent` value is used until the page is reloaded.

File: packages/webview_flutter/lib/src/webview_method_channel.dart [hunk 6]
```diff
@@ -122,6 +127,7 @@ class MethodChannelWebViewPlatform implements WebViewPlatformController {
     _addIfNonNull('jsMode', settings.javascriptMode?.index);
     _addIfNonNull('hasNavigationDelegate', settings.hasNavigationDelegate);
     _addIfNonNull('debuggingEnabled', settings.debuggingEnabled);
+    _addIfNonNull('userAgent', settings.userAgent);
```
Comment (amirh) [inline packages/webview_flutter/lib/src/webview_method_channel.dart]: What happens if we rebuild a WebView that had an explicitly set `userAgent` and set the `userAgent` to null. IIUC the way this currently propagates will lead to preserving the previous user agent. I'd expect it to go back to the platform's default user agent.

We should add an test for this scenario as well (build with an explicitly set user agent, rebuild with user agent set to null)

File: packages/webview_flutter/lib/platform_interface.dart [hunk 7]
```diff
@@ -152,6 +152,11 @@ abstract class WebViewPlatformController {
     throw UnimplementedError(
         "WebView removeJavascriptChannels is not implemented on the current platform");
   }
+
+  Future<String> getUserAgent() {
```
Comment (amirh) [inline packages/webview_flutter/lib/platform_interface.dart]: Add Dartdoc

File: packages/webview_flutter/lib/platform_interface.dart [hunk 8]
```diff
@@ -176,16 +182,24 @@ class WebSettings {
   /// See also: [WebView.debuggingEnabled].
   final bool debuggingEnabled;
 
+  /// The User Agent used for all following requests
```
Comment (amirh) [inline packages/webview_flutter/lib/platform_interface.dart]: See Dartdoc comments on the userAgent field in WebView

File: packages/webview_flutter/ios/Classes/FlutterWebView.m [hunk 9]
```diff
@@ -321,4 +326,29 @@ - (void)registerJavaScriptChannels:(NSSet*)channelNames
...
+           result(userAgent);
+         }
+       }];
+}
+
+- (void)updateUserAgent:(NSString*)userAgent {
+  if (@available(iOS 9.0, *)) {
+    [_webView setCustomUserAgent:userAgent];
+  } else {
+    NSLog(@"Updating UserAgent is not supported for Flutter WebViews prior to iOS 9.");
```
Comment (amirh) [inline packages/webview_flutter/ios/Classes/FlutterWebView.m]: Mention that in the Dartdoc for WebView.userAgent.

File: packages/webview_flutter/lib/src/webview_method_channel.dart (hunk 6 above)
Comment (LuisThein) [reply packages/webview_flutter/lib/src/webview_method_channel.dart]: Makes sense. I changed it so that the user agent can be set to null and the default platform user agent is used after a rebuild. Additionally I adapted the same logic for the FakePlatformWebView unit tests.

**Comment by LuisThein** [comment]
Added requested changes and updated branch.

File: packages/webview_flutter/lib/src/webview_method_channel.dart (hunk 6 above)
Comment (amirh) [reply packages/webview_flutter/lib/src/webview_method_channel.dart]: Thanks!
I think we should avoid updating the user agent each time the widget is rebuilt.
It seems like we're lacking the ability to express presence/absence of a setting that may be null. One way to do it would be to use some kind of an optional type.

I pushed https://github.com/flutter/plugins/pull/1920/commits/79761515b33c3e96234e7b0555651c74f971adee with a proposal on how to do that, let me know whether that makes sense to you.

File: packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java [hunk 10]
```diff
@@ -64,6 +64,11 @@
       String url = (String) params.get("initialUrl");
       webView.loadUrl(url);
     }
+
+    if (params.containsKey("userAgent")) {
```
Comment (amirh) [inline packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java]: This should be done before `loadUrl` for the initial URL.

Comment (amirh) [reply packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java]: pushed fix

**Comment by amirh** [comment]
@LuisThein I pushed a few changes to your branch.
I'm ok with the current state of this PR.
Let me know whether the updates I pushed makes sense to you as well.

File: packages/webview_flutter/lib/src/webview_method_channel.dart (hunk 6 above)
Comment (LuisThein) [reply packages/webview_flutter/lib/src/webview_method_channel.dart]: I think that is a nice way to handle nullable settings and I agree that updating the user agent every rebuild is unnecessary.

File: packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java (hunk 10 above)
Comment (LuisThein) [reply packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java]: Makes sense, thanks for fixing it.

File: packages/webview_flutter/ios/Classes/FlutterWebView.m [hunk 11]
```diff
@@ -347,4 +352,29 @@ - (void)registerJavaScriptChannels:(NSSet*)channelNames
   }
 }
 
+- (void)onGetUserAgent:(FlutterMethodCall*)call result:(FlutterResult)result {
+  [_webView
+      evaluateJavaScript:@"navigator.userAgent"
```
Comment (amirh) [inline packages/webview_flutter/ios/Classes/FlutterWebView.m]: I just realized we can't rely on JavaScript here as it may be disabled.
Sorry for missing this in previous iterations.

I guess if WkWebView doesn't support getting the user agent we should probably let our users inject js to get it as they are the ones in control of the js policy.

My suggestion for this PR would be to not include the `getUserAgent` method, and inject js to get the user agent in the e2e tests.

We should file an issue for `getUserAgent` if that turns out to be important to many users we should consider adding it as an Android only API (maybe as part of the Android controller).

Comment (LuisThein) [reply packages/webview_flutter/ios/Classes/FlutterWebView.m]: I see, it makes sense that we shouldn't rely on injected javascript to get the user agent in the WkWebView. As the main goal of this pr was to add the ability to set the user agent, I don't mind removing the `getUserAgent` method. I removed all methods involved in getting the user agent and updated the tests accordingly.

File: packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java [hunk 12]
```diff
@@ -72,21 +76,29 @@ public View getView() {
   }
 
   // @Override
-  // This is overriding a method that hasn't rolled into stable Flutter yet. Including the
-  // annotation would cause compile time failures in versions of Flutter too old to include the new
-  // method. However leaving it raw like this means that the method will be ignored in old versions
+  // This is overriding a method that hasn't rolled into stable Flutter yet.
```
Comment (amirh) [inline packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java]: Please revert unrelated re-formatting

File: packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java [hunk 13]
```diff
@@ -72,21 +76,29 @@ public View getView() {
...
+  // stable.
   public void onInputConnectionUnlocked() {
     webView.unlockInputConnection();
   }
 
   // @Override
-  // This is overriding a method that hasn't rolled into stable Flutter yet. Including the
-  // annotation would cause compile time failures in versions of Flutter too old to include the new
-  // method. However leaving it raw like this means that the method will be ignored in old versions
+  // This is overriding a method that hasn't rolled into stable Flutter yet.
```
Comment (amirh) [inline packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java]: Also here

File: packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java [hunk 14]
```diff
@@ -261,7 +276,8 @@ private void updateJsMode(int mode) {
   }
 
   private void updateAutoMediaPlaybackPolicy(int mode) {
-    // This is the index of the AutoMediaPlaybackPolicy enum, index 1 is always_allow, for all
+    // This is the index of the AutoMediaPlaybackPolicy enum, index 1 is
```
Comment (amirh) [inline packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java]: Same here.

File: packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java (hunk 12 above)
Comment (LuisThein) [reply packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java]: Oops, should be fixed now.
